racket-common.el

```diff
@@ -628,15 +631,78 @@ This function is a suitable element for the list variable
+The defined name is the second regexp group.")
+
 (defun racket--beginning-of-defun-function ()
-  "Like `beginning-of-defun' but aware of Racket module forms."
+  "Move point backward to the beginning of a definition.
+The code moves only to the next inner most definition, if nested
+defines are present. If the point is already on a definition
+which is not a toplevel one, it will move to the next outer
+definition, thus repeated calls of this function will move up on
+the nested definitions."
   (let ((orig (point)))
-    (racket--escape-string-or-comment)
-    (pcase (racket--module-level-form-start)
-      (`() (ignore-errors (backward-sexp 1)))
-      (pos (goto-char pos)))
+    (catch 'found
+      ;; If we are after the close paren of a definition, but before
+      ;; the next definition, move to the definition start...
+      (ignore-errors (backward-sexp))
+      (when (looking-at racket--defun-start-rx)
+        (throw 'found (point)))
+      (goto-char orig)
+      (let ((parens (reverse (nth 9 (syntax-ppss (point))))))
+        ;; If we are inside a nested sexp, try to move to the
+        ;; innermost definition...
+        (dolist (pos parens)
+          (goto-char pos)
+          (when (looking-at racket--defun-start-rx)
+            (throw 'found pos)))
+        ;; ... we could not find a definition start, move to top level
+        ;; and start scanning sexps backwards until we find a
+        ;; definition, or reach the beginning of the buffer.
+        (unless (eq parens nil)
+          (goto-char (car (last parens))))
+        (backward-sexp 1)
+        (while (not (or (bobp) (looking-at racket--defun-start-rx)))
+          (backward-sexp 1))))
+    (goto-char (or (racket--sexp-comment-start (point)) (point)))
     (/= orig (point))))
 
+(defun racket--end-of-defun-function ()
+  "Move point to the end of the current function definition.
+This is intended to be used as the `end-of-defun-function' value,
+and assumes point is set up correctly."
+  (let ((orig (point)))
+    (condition-case err
+        (forward-sexp 1)
+      (scan-error
+       ;; scan errors happen if we are inside a nested definition and
+       ;; move to the end.  Go outside the nested definition in this
+       ;; case.
+       (let ((parens (nth 9 (syntax-ppss orig))))
+         (if (eq parens nil)
+             (goto-char (point-max))
+           (progn
+             (goto-char (car parens))
+             (forward-sexp 1))))))))
+
+(defun racket--current-defun-name ()
+  "Return the current function name.
+
+If the point is inside a nested function definition, return a
+string that represents the concatenation of the nested function names."
+  (save-excursion
+    (unless (looking-at racket--defun-start-rx)
+      (racket--beginning-of-defun-function))
+    (when (looking-at racket--defun-start-rx)
+      (let ((name (list (match-string-no-properties 2)))
+            (parens (reverse (nth 9 (syntax-ppss (point))))))
```

Comment:
To make code a little more self-documenting, I added `racket-ppss.el` with accessor aliases. But it only goes up to `(nth 8)`. Perhaps you could add one for `(nth 9)` named something like `racket--ppss-open-paren-positions`, and use it here? (If that seems like too much busywork for you, no worries; I might come back and do it later myself.)